doc/adr/0010-atomic-composite-grapth-operations.md

```diff
@@ -0,0 +1,267 @@
+    - ((⚛️**`:page/merge`**))
+    - ((⚛️**`:page/remove`**))
+- *Block Ops*
+    - ((⚛️**`:block/new`**))
+    - ((⚛️**`:block/save`**))
+    - ((⚛️**`:block/open`**))
+    - ((⚛️**`:block/move`** ⭐️))
+    - ⚛️**`:block/remove`**
+- *Shortcut Ops*
+    - (({{[[TODO]]}} ⚛️**`:shortcut/new`**))
+    - (({{[[TODO]]}} ⚛️**`:shortcut/remove`**))
+    - (({{[[TODO]]}} ⚛️**`:shortcut/move`**))
+
+
+((⎄ Composite Graph Ops)) are a way to group ((⚛️ Atomic Graph Ops)).
+
+We'll only need one type of Composite Graph Ops, that is Consequence Event.
+
+Consequence Event is 2 things:
+- Trigger Event (like `:paste`)
+- Consequence Events List
+    - Containing however many events to include into transaction resolution.
+    - These events can be more of Consequence Events, or ((⚛️ Atomic Graph Ops)).
+
+
+This will allow for easier Temporality,
+We'll be able to represent Main Action that moved Knowledge Graph forward.
+
+It will maintain information about `:paste` events that injected sub-tree with ((⚛️ Atomic Graph Ops)).
+
+## Consequences
+
+We'll have additional work of Porting Semantic Events to use Atomic Graph Ops and compositions.
+
+We have working `:paste` & `:block/save` events, they are broken not ported now.
+
+We'll have smaller amount of tests in order to provide correctness guarantees
+
+
+## Additional Resources
+
+### Catalog of operations
+
+- Types of Graph Operations
+    - ⚛️ Atomic Graph Ops
+        - Not divisible Graph Ops
+        - Operations like create new block, create page, save block
+    - ⎄ Composite Graph Ops
+        - Collection of events to be executed on the graph
+        - Like `:block/save` when new link is discovered, should produce also `:page/create` event
+
+- List of Atomic Graph Operations:
+    - Page Ops
+        - ⚛️**`:page/new`**
+            - ((⚛️ Atomic Graph Ops))
+            - *Input*
+                - `title` - Page title page to be created
+                - `page-uid` - `:block/uid` of page to be created
+                - `block-uid` - `:block/uid` of 1st block to be created in page to be created
+        - ⚛️**`:page/rename`**
+            - ((⚛️ Atomic Graph Ops))
+            - *Input*
+                - `page-uid` - `:block/uid` of page to be renamed
+                - `new-name` - Page should have this name after operation
+                - `old-name` - ^^To Remove^^ This is accidental, and shouldn't be provided
+        - ⚛️**`:page/merge`**
+            - ((⚛️ Atomic Graph Ops))
+            - *Input*
+                - `page-uid` - `:block/uid` of page to be merged into `new-page`
+                - `new-page` - page name of a page we'll merge contents of `page-uid` page into
+                - `old-name` - ^^To Remove^^ This is accidental, and shouldn't be provided
+        - ⚛️**`:page/remove`**
+            - ((⚛️ Atomic Graph Ops))
+            - *Input*
+                - `page-uid` - `:block/uid` of the page to be deleted
+    - Block Ops
+        - ⚛️**`:block/new`**
```

Comment:
Block new, as defined here, is the same as block/move WRT the args. I imagine the difference is that it also sets an empty string or something. But then it's a composite between save and move.

This distinction is meaningful is conflict resolution. You can succeed at editing the content but fail on setting the position, and other combinations. Composing these two operations also ensures idempotency.

Reply:
How can you save or move block that doesn't exist?
IMHO Block creation is atomic, and isn't composition of save and move.

![](https://media.giphy.com/media/SwkQVEYIawjkVJE2jI/giphy.gif)